### ISSUE TITLE
`WaitForNonStaleProjectionDataAsync` does not fetch projection progress correctly

### ISSUE THREAD

**Issue by baltie** [JasperFx/marten#3080]
We've been trying to use the new `IDocumentStore.WaitForNonStaleProjectionDataAsync()` test extension method in our integration tests to wait for the async projections to stabilize before moving on in the test, using code that depends on the projections being up to date.

However, it seems `WaitForNonStaleProjectionDataAsync` doesn't work correctly yet. Reading the source code, the projection progress is retrieved only once, and then it will keep checking that same result over and over until timeout. See https://github.com/JasperFx/marten/blob/61de7fcbe5450d42e401580fa76bb915a2e49430/src/Marten/Events/AsyncProjectionTestingExtensions.cs#L67-L72

Here, `AllProjectionProgress` returns an `IReadOnlyList<ShardState>` and fetching the projection progress will not be executed again. All our tests are thus cancelled after the timeout. I've verified in the database that the async projections have run, but it is not picked up by `WaitForNonStaleProjectionDataAsync`, and I strongly suspect it is due to the above code snippet.

**Comment by baltie** [comment]
See #2624. There's the additional problem of `WaitForNonStaleProjectionDataAsync` not waiting for all async projections to appear in `mt_event_progression` before proceeding. This only occurs when starting with an empty database, but that is the case when we run our integration tests - we let Marten create everything.

**Comment by jeremydmiller** [comment]
Pfft, this is the 3rd or 4th attempt to improve this thing. I have a reproduction project from a JasperFx client that I'm about to use to try to improve this. This is the next thing up for Marten at least.

**Comment by a-shtifanov-laya** [comment]
i'm actually preparing a PR for this 😃

**Comment by jeremydmiller** [comment]
Closed by https://github.com/JasperFx/marten/commit/96a9e2032d9fc44fafdaacedae6d399625cc6bc1

**Comment by daveHylde** [comment]
Sorry to bring this back to light again, but I'm facing these issues still, also after upgrading to v7.18.0.
The `WaitForNonStaleProjectionDataAsync` call works for a single test in an `IClassFixture`, but the rest of the tests gets the timeout excetion. All pass when run one at a time.
@baltie can you confirm that you don't have this issue anymore?

**Comment by daveHylde** [comment]
I've also noticed that this only is an issue when initializing tests with `ResetAllData`.

**Comment by jeremydmiller** [comment]
Hey folks, please don't add comments on an already closed issue. I'll never see these unless someone brings it to my attention. Next time just open a brand new issue and link it to the old if you'd like.
